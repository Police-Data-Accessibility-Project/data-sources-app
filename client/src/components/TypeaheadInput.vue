<!-- TODO: Once this component is sufficiently generic, move to design-system? -->
<template>
	<div
		:id="wrapperId"
		data-test="typeahead-wrapper"
		class="pdap-typeahead"
		:class="{ 'pdap-typeahead-expanded': isListOpen }"
	>
		<label v-if="$slots.label" class="col-span-2" :for="id">
			<slot name="label" />
		</label>

		<div v-if="$slots.error && error" class="pdap-input-error-message">
			<!-- TODO: aria-aware error handling?? Not just here but in other input components as well? -->
			<slot name="error" />
		</div>
		<div v-else-if="error" class="pdap-input-error-message">{{ error }}</div>

		<input
			:id="id"
			ref="inputRef"
			v-model="input"
			data-test="typeahead-input"
			class="pdap-typeahead-input"
			type="text"
			:placeholder="placeholder"
			autocomplete="off"
			v-bind="$attrs"
			@input="onInput"
			@focus="onFocus"
			@blur="onBlur"
			@keydown.down.prevent="onArrowDown"
		/>
		<ul
			v-if="itemsToDisplay?.length && inputRef?.value"
			data-test="typeahead-list"
			class="pdap-typeahead-list"
		>
			<li
				v-for="(item, index) in itemsToDisplay"
				:key="index"
				class="pdap-typeahead-list-item"
				data-test="typeahead-list-item"
				role="button"
				tabindex="0"
				@click="selectItem(item)"
				@keydown.enter.prevent="selectItem(item)"
				@keydown.down.prevent="onArrowDown"
				@keydown.up.prevent="onArrowUp"
			>
				<slot v-if="$slots.item" name="item" v-bind="item" />
				<span v-else>{{ boldMatchText(item) }}</span>
			</li>
		</ul>
		<ul
			v-else-if="typeof itemsToDisplay === 'undefined' && input.length > 1"
			class="pdap-typeahead-list"
			data-test="typeahead-list-not-found"
		>
			<li class="max-w-[unset]">
				<slot
					v-if="$slots['not-found']"
					name="not-found"
					v-bind="notFound ?? {}"
				/>
				<span v-else>
					<strong>No results found.</strong> Please check your spelling.
				</span>
			</li>
		</ul>
	</div>
</template>

<script setup>
import {
	ref,
	computed,
	onMounted,
	onUnmounted,
	watch,
	onBeforeUpdate,
} from 'vue';

/* Props and emits */
const props = defineProps({
	id: {
		type: String,
	},
	placeholder: {
		type: String,
		default: '',
	},
	items: {
		type: Array,
	},
	formatItemForDisplay: {
		type: Function,
	},
	error: {
		type: String,
	},
});
const emit = defineEmits(['onInput', 'onFocus', 'onBlur', 'selectItem']);

/* Refs and reactive vars */
const inputRef = ref();
const input = ref('');

/* Computed vars */
const wrapperId = computed(() => `${props.id}_wrapper`);
const itemsToDisplay = computed(() => props.items);
const isListOpen = computed(
	() =>
		(itemsToDisplay.value?.length && inputRef?.value?.value) ||
		(typeof itemsToDisplay.value === 'undefined' && input.value.length > 1),
);

/* Lifecycle methods and listeners */
onMounted(() => {
	window.addEventListener('resize', setInputPositionForList);
});

onBeforeUpdate(setInputPositionForList);

onUnmounted(() => {
	window.removeEventListener('resize', setInputPositionForList);
});

/* Watch expressions */
watch(
	() => inputRef.value,
	(ref) => {
		if (ref) setInputPositionForList();
	},
);

/* Methods */
function setInputPositionForList() {
	document.documentElement.style.setProperty(
		'--typeaheadBottom',
		inputRef.value.offsetTop + inputRef.value.offsetHeight + 'px',
	);
	document.documentElement.style.setProperty(
		'--typeaheadListWidth',
		inputRef.value.offsetWidth + 'px',
	);
}
function onInput(e) {
	emit('onInput', e);
}
function onFocus(e) {
	if (Array.isArray(itemsToDisplay.value) && !itemsToDisplay.value.length) {
		clearInput();
		emit('selectItem', undefined);
	}

	emit('onFocus', e);
}
function onBlur(e) {
	emit('onBlur', e);
}

function onArrowDown() {
	const items = Array.from(
		document.getElementsByClassName('pdap-typeahead-list-item'),
	);

	const focusedIndex = items.indexOf(document.activeElement);

	if (focusedIndex === items.length - 1) return;

	if (focusedIndex === -1) {
		items[0].focus();
	} else {
		items[focusedIndex + 1].focus();
	}
}

function onArrowUp() {
	const items = Array.from(
		document.getElementsByClassName('pdap-typeahead-list-item'),
	);

	const focusedIndex = items.indexOf(document.activeElement);

	if (focusedIndex === 0) {
		inputRef.value.focus();
	} else {
		items[focusedIndex - 1].focus();
	}
}

function selectItem(item) {
	input.value = props.formatItemForDisplay
		? props.formatItemForDisplay(item)
		: item;
	inputRef.value.blur();
	emit('selectItem', item);
}

function escapeRegExp(string) {
	return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
function boldMatchText(text) {
	const regexp = new RegExp(`(${escapeRegExp(input.value)})`, 'ig');
	return text.replace(regexp, '<strong>$1</strong>');
}
function clearInput() {
	input.value = '';
}
// function getInput() {
// 	return inputRef.value;
// }
function focusInput() {
	inputRef.value.focus();
	onFocus();
}
// function blurInput() {
// 	inputRef.value.blur();
// 	onBlur();
// }

defineExpose({
	boldMatchText,
	clearInput,
	focusInput,
	get value() {
		return input.value;
	},
});
</script>

<style>
.pdap-typeahead {
	@apply leading-normal w-full flex flex-col;
}

.pdap-typeahead label {
	@apply max-w-[max-content] text-lg py-1 font-medium;
}

.pdap-typeahead-input {
	@apply rounded-md;
}

.pdap-typeahead-input,
.pdap-typeahead-list {
	@apply bg-neutral-50 border-2 border-neutral-500 border-solid p-2 text-neutral-950;
}

.pdap-typeahead-input::placeholder {
	@apply text-neutral-600 text-lg;
}

.pdap-typeahead-input:focus,
.pdap-typeahead-input:focus-within,
.pdap-typeahead-input:focus-visible {
	@apply border-2 border-brand-gold border-solid outline-none;
}

.pdap-typeahead-list {
	@apply absolute w-[var(--typeaheadListWidth)] top-[var(--typeaheadBottom)] z-50 overflow-scroll;
}

.pdap-typeahead-list-item {
	@apply mt-1 max-w-[unset] p-2 flex items-center gap-6 text-sm @md:text-lg;
}

.pdap-typeahead-list-item:focus,
.pdap-typeahead-list-item:focus-visible,
.pdap-typeahead-list-item:focus-within,
.pdap-typeahead-list-item:hover {
	@apply outline-none text-neutral-950 bg-neutral-300;
}
</style>
