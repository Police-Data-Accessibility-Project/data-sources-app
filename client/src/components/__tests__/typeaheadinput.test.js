import TypeaheadInput from '@/components/TypeaheadInput.vue';
import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import { nextTick } from 'vue';

const MOCK_SEARCH_VALUE = 'lev';
const MOCK_ITEMS = ['Levantine', 'Leviathan', 'Levi', 'Levant'];
// To do: update test to handle more complex items
// const MOCK_ITEMS = [
// 	{
// 		county: 'Hockley',
// 		display_name: 'Levelland',
// 		locality: 'Levelland',
// 		state: 'Texas',
// 		type: 'Locality',
// 	},
// 	{
// 		county: 'Franklin',
// 		display_name: 'Leverett',
// 		locality: 'Leverett',
// 		state: 'Massachusetts',
// 		type: 'Locality',
// 	},
// 	{
// 		county: 'Levy',
// 		display_name: 'Levy',
// 		locality: null,
// 		state: 'Florida',
// 		type: 'County',
// 	},
// 	{
// 		county: 'Marion',
// 		display_name: 'Belleview',
// 		locality: 'Belleview',
// 		state: 'Florida',
// 		type: 'Locality',
// 	},
// ];

// Setup function to mount the component with optional props
const mountComponent = (props = {}) => {
	return mount(TypeaheadInput, {
		props: {
			items: MOCK_ITEMS,
			...props,
		},
		attachTo: document.body,
	});
};

// Function to get the input element
const getInput = (wrapper) => wrapper.find('[data-test="typeahead-input"]');

// Function to get the list items
const getListItems = (wrapper) =>
	wrapper.findAll('[data-test="typeahead-list-item"]');

// Function to set input value and wait for update
const setInputValue = async (input, value) => {
	input.setValue(value);
	await nextTick();
};

// Function to focus input and set value
const focusAndSetInput = async (wrapper, value) => {
	const input = getInput(wrapper);
	await input.trigger('focus');
	await setInputValue(input, value);
};

// Test suite
describe('TypeaheadInput', () => {
	it('emits onInput event', async () => {
		const wrapper = mountComponent();
		const input = getInput(wrapper);

		await input.trigger('input');

		expect(wrapper.html()).toMatchSnapshot();
		expect(wrapper.emitted().onInput).toBeTruthy();
	});

	it('emits onFocus event and clears input if no items', async () => {
		const wrapper = mountComponent();
		const input = getInput(wrapper);
		await setInputValue(input, MOCK_SEARCH_VALUE);

		wrapper.setProps({ items: [] });
		await input.trigger('click');
		await input.trigger('focus');

		await nextTick();

		expect(wrapper.html()).toMatchSnapshot();
		expect(wrapper.emitted().onFocus).toBeTruthy();
		expect(input.text()).toBe('');
	});

	it('emits onBlur event', async () => {
		const wrapper = mountComponent();
		const input = getInput(wrapper);

		await input.trigger('blur');

		expect(wrapper.html()).toMatchSnapshot();
		expect(wrapper.emitted().onBlur).toBeTruthy();
	});

	it('focuses next list item on arrow down', async () => {
		const wrapper = mountComponent();
		await focusAndSetInput(wrapper, MOCK_SEARCH_VALUE);
		const [item1, item2] = getListItems(wrapper);

		await getInput(wrapper).trigger('click');
		await getInput(wrapper).trigger('keydown', {
			key: 'ArrowDown',
			keyCode: 40,
		});

		expect(document.activeElement).toBe(item1.element);

		await item1.trigger('keydown', { key: 'ArrowDown', keyCode: 40 });

		expect(document.activeElement).toBe(item2.element);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('focuses previous list item on arrow up', async () => {
		const wrapper = mountComponent();
		await focusAndSetInput(wrapper, MOCK_SEARCH_VALUE);
		const [item1, item2] = getListItems(wrapper);

		await item2.trigger('focus');
		await item2.trigger('keydown', { key: 'ArrowUp', keyCode: 38 });

		expect(document.activeElement.innerText).toContain(
			item1.element.textContent,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('focuses input on arrow up from first list item', async () => {
		const wrapper = mountComponent();
		await focusAndSetInput(wrapper, MOCK_SEARCH_VALUE);
		const [item1] = getListItems(wrapper);

		await item1.trigger('focus');
		await item1.trigger('keydown', { key: 'ArrowUp', keyCode: 38 });

		expect(document.activeElement).toBe(getInput(wrapper).element);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('selects item and emits selectItem event on click', async () => {
		const wrapper = mountComponent();
		await focusAndSetInput(wrapper, MOCK_SEARCH_VALUE);
		const item1 = getListItems(wrapper)[0];

		await item1.trigger('click');
		expect(wrapper.emitted().selectItem).toBeTruthy();
		expect(wrapper.emitted().selectItem[0][0]).toEqual(MOCK_ITEMS[0]);
	});

	it('selects item and emits selectItem event on enter', async () => {
		const wrapper = mountComponent();
		await focusAndSetInput(wrapper, MOCK_SEARCH_VALUE);
		const item1 = getListItems(wrapper)[0];

		await item1.trigger('keydown', { key: 'Enter', keyCode: 13 });
		expect(wrapper.emitted().selectItem).toBeTruthy();
		expect(wrapper.emitted().selectItem[0][0]).toEqual(MOCK_ITEMS[0]);
	});

	// TODO: set up test as above for complex items
	// it('formats text correctly for different item types', () => {
	// 	const { vm } = mountComponent();

	// 	const localityItem = {
	// 		display_name: 'City',
	// 		county: 'County',
	// 		state: 'California',
	// 		type: 'Locality',
	// 	};
	// 	const countyItem = {
	// 		display_name: 'County',
	// 		state: 'California',
	// 		type: 'County',
	// 	};
	// 	const stateItem = { display_name: 'California', type: 'State' };

	// 	expect(vm.formatText(localityItem)).toBe('City County CA');
	// 	expect(vm.formatText(countyItem)).toBe('County CA');
	// 	expect(vm.formatText(stateItem)).toBe('California');
	// });

	it('bolds matched text correctly', () => {
		const { vm } = mountComponent();
		vm.input = 'test';

		expect(vm.boldMatchText('This is a test string')).toBe(
			'This is a <strong>test</strong> string',
		);
	});

	it('computes wrapperId correctly', () => {
		const { vm } = mountComponent({
			id: 'test-id',
		});

		expect(vm.wrapperId).toBe('test-id_wrapper');
	});

	it('computes itemsToDisplay correctly', () => {
		const { vm } = mountComponent();

		expect(vm.itemsToDisplay).toEqual(MOCK_ITEMS);
	});
});
