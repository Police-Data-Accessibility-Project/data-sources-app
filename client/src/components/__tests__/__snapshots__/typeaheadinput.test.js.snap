// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeaheadInput > emits onBlur event 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <!--v-if-->
</div>
`;

exports[`TypeaheadInput > emits onFocus event and clears input if no items 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <!--v-if-->
</div>
`;

exports[`TypeaheadInput > emits onInput event 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <!--v-if-->
</div>
`;

exports[`TypeaheadInput > focuses input on arrow up from first list item 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <ul class="pdap-typeahead-list">
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>antine
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>iathan
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>i
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>ant
      </span>
    </li>
  </ul>
</div>
`;

exports[`TypeaheadInput > focuses next list item on arrow down 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <ul class="pdap-typeahead-list">
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>antine
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>iathan
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>i
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>ant
      </span>
    </li>
  </ul>
</div>
`;

exports[`TypeaheadInput > focuses previous list item on arrow up 1`] = `
<div class="pdap-typeahead" id="undefined_wrapper">
  <!--v-if-->
  <!--v-if-->
  <input autocomplete="off" class="pdap-typeahead-input" placeholder type="text">
  <ul class="pdap-typeahead-list">
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>antine
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>iathan
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>i
      </span>
    </li>
    <li class="pdap-typeahead-list-item" role="button" tabindex="0">
      <span>
        <strong>Lev</strong>ant
      </span>
    </li>
  </ul>
</div>
`;
